### PR TITLE
Feature/no af networking dependency

### PR DESCRIPTION
**Pull request by BenMaer**
Removed 'AFOAuth2Manager' and 'AFNetworking' dependencies. Replaced with NSUrlSession code.

### CODE REVIEW

File: Pod/Classes/API/ABMAuth2Manager.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+
+@class ABMAuthenticationCredentials;
+
+
+
+
+
+typedef void(^abm_Auth2Manager_authenticationSuccessBlock) (ABMAuthenticationCredentials *credentials);
```
Comment (andresbrun) [inline Pod/Classes/API/ABMAuth2Manager.h]: Maybe there is no need of using underscore notation. I would use just `Auth2ManagerAuthenticationSuccessBlock`. And the same for successBlock and failureBlock.

File: Pod/Classes/API/ABMAuth2Manager.m [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+
+
+NSString * const kABMAuth2Manager_AuthCodeGrantType = @"authorization_code";
+
+
+
+
+
+typedef NS_ENUM(NSInteger, ABMAuth2Manager_HTTPMethodType) {
+	ABMAuth2Manager_HTTPMethodType_GET,
```
Comment (andresbrun) [inline Pod/Classes/API/ABMAuth2Manager.m]: Same here:
`ABMAuth2ManagerHTTPMethodTypeGET`
`ABMAuth2ManagerHTTPMethodTypePOST`
`ABMAuth2ManagerHTTPMethodTypePUT`

File: Pod/Classes/API/ABMAuth2Manager.m [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+}
+
+#pragma mark - Authenticate
+- (NSURLSessionDataTask *)authenticateUsingOAuthWithURLString:(NSString *)URLString
+														 code:(NSString *)code
+												  redirectURI:(NSString *)uri
+													  success:(abm_Auth2Manager_authenticationSuccessBlock)success
+													  failure:(abm_Auth2Manager_failureBlock)failure
+{
+	if (URLString.length == 0)
```
Comment (andresbrun) [inline Pod/Classes/API/ABMAuth2Manager.m]: Code style:

```
if () {
}
```

File: Pod/Classes/API/SoundCloudPort.m [hunk 4]
```diff
@@ -79,7 +93,9 @@ - (void)handleInvalidTokenWithResult:(void (^)(BOOL success))resultBlock {
 }
 
 - (BOOL)isValidToken {
-    return self.credentials && ![self.credentials isExpired];
+	ABMAuthenticationCredentials* credentials = self.credentials;
+    return ((credentials != nil) &&
```
Comment (andresbrun) [inline Pod/Classes/API/SoundCloudPort.m:98]: You don't need so many parentesis here:

```
return credentials && ![credentials isExpired];
```

File: Pod/Classes/Controllers/SoundCloudLoginWebViewController.m [hunk 5]
```diff
@@ -58,7 +58,19 @@ - (NSString *)extractCodeFromRequest:(NSURLRequest *)request {
 }
 
 - (BOOL)isFlowFinishWithRequest:(NSURLRequest *)request {
-    NSString *baseURL = [@[request.URL.scheme, request.URL.host] componentsJoinedByString:@"://"];
+	NSString* scheme = request.URL.scheme;
+	if (scheme == nil)
```
Comment (andresbrun) [inline Pod/Classes/Controllers/SoundCloudLoginWebViewController.m:62]: Code style

**Comment by andresbrun** [comment]
So finally we are using parts of AFOAuth2Manager to accomplish authentication, right? I am thinking that another possibility to avoid dependency conflicts would be using submodules.

Adding AFOAuth2 and AFNetworking/AFHTTPRequestOperation as submodules in this project we can remove those dependencies, but I've never tried it :( Idk if there will be any compilation issue because there are files with same name.

http://git-scm.com/book/en/v2/Git-Tools-Submodules

**Comment by BenMaer** [comment]
We are no longer using AFOAuth2Manager directly. I've created code which in many ways duplicates its functionality, but uses NSURLSession instead.

I don't think there's any reason to keep 'AFOAuth2Manager' as a submodule. The whole point of this branch was to remove the need for 'AFOAuth2Manager' entirely, since it depends on AFNetworking, and other libraries (like Restkit) could have conflict with this library.

**Comment by andresbrun** [comment]
The whole points of this branch is to remove AFNetworking as a dependency, which maybe doesn't imply to remove it from the project. Using code of well maintain 3-parties libraries save us the effort of keeping it up to date. So I think is better, if there is a way, to use AFNetworking (or any other connection library that supports oauth2) instead of replicate part of the functionality because in the end is gonna be more code to maintain.

If there is no way to do it I will take this approach but it has to have the same functionality than the current one (every already implemented request).

**Comment by BenMaer** [comment]
Okay, I'll go ahead and implement the rest of the requests. Should we close this pull request? I will make another once I've implemented the rest of the requests. I'll also make another library for the networking code itself, making it easier to maintain. Hopefully that be enough to remove AFNetworking entirely.

**Comment by andresbrun** [comment]
You can push new commit in your branch and they will reflex here, you don't need to create another PR ;)

**Comment by BenMaer** [comment]
This has been tested with your sample project's UI. Note that fetch playlists isn't working, but it wasn't working on the current master either. What does work is authentication, search songs, and get song.

**Comment by andresbrun** [comment]
Just out of curiosity, why did you remove `ABMSoundCloudAPI.xcworkspace`? it is needed to run the example properly.

**Comment by BenMaer** [comment]
Dang, I did not mean to do that... I'm not even sure how it happened. My mistake, I've added it back in.

**Comment by andresbrun** [comment]
Ok, it seems to work properly. I am gonna accept the PR but I wanna point somethings out:
- In the end you change `AFNetworking` by this `NSURLSession-Resplendent`. Which is less maintain and less supported.
- We've loose download song feature, which it could be very interesting.
- The code style hasn't been kept, anyway I will fix it since is one of my habits.
  But overall: thanks for the PR :+1:

**Comment by BenMaer** [comment]
Thanks for the notes. You're right that `NSURLSession-Resplendent` won't be as maintained/supported as `AFNetworking`, but I'll be using `NSURLSession-Resplendent` in a few of my projects, so it won't be entirely sitting in the dust :) something worth pointing out is `NSURLSession-Resplendent` has way less complexity than `AFNetworking`, and uses a lot more native components, so it's likely this library won't need the amount of support and maintenance `AFNetworking` has.

Also, please don't hesitate to request any changes to `NSURLSession-Resplendent`, either through github issue(s) or pull requests.

**Pull request merged**